### PR TITLE
[CDEC-313] regression tests for subscription

### PR DESCRIPTION
**Pull request by avieth**
## Description

Ensure proper exception handling properties of `networkSubscribeTo`.

This also introduces test suite infrastructure to cardano-sl-infra.

## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [ ] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by avieth** [comment]
@erikd Did I do the test suite format right?

File: infra/Pos/Diffusion/Subscription/Common.hs [hunk 1]
```diff
@@ -238,15 +206,57 @@ networkSubscribeTo' logTrace oq bucket nodeType peersVar keepalive status sendAc
...
-    alterPeerSubStatus :: NodeId -> Maybe SubscriptionStatus -> IO ()
-    alterPeerSubStatus peer s = atomically $ do
+    alterPeerSubStatus :: Maybe SubscriptionStatus -> IO ()
+    alterPeerSubStatus s = atomically $ do
         stats <- readTVar status
         let !stats' = Map.alter fn peer stats
         writeTVar status stats'
         where
             fn :: Maybe SubscriptionStatus -> Maybe SubscriptionStatus
             fn x = getOption (Option x <> Option s)
```
Comment (coot) [inline infra/Pos/Diffusion/Subscription/Common.hs]: I wonder if we're using the right monoid here, `Option Nothing` is a unit for `Option` so `alterPeerSubStatus Nothing` will not remove the peer from the `TVar`.

Comment (avieth) [reply infra/Pos/Diffusion/Subscription/Common.hs]: Oh right this is per-peer, so assuming there's only one writer per-peer then we can just overwrite the old status.

File: infra/test/Spec.hs [hunk 2]
```diff
@@ -0,0 +1,9 @@
+module Spec
+    ( spec
+    ) where
+
+import           Test.Hspec
+import qualified Test.Pos.Infra.Diffusion.Subscription.SubscriptionSpec (spec)
+
+spec :: Spec
+spec = Test.Pos.Infra.Diffusion.Subscription.SubscriptionSpec.spec
```
[the remark this answers is not in the document]
Comment (avieth) [reply infra/test/Spec.hs]: I've made another test, for subscription status, so once I push it this file will define more than one spec. Isn't using a `Spec.hs` file the idiomatic way?

**Pull request merged**